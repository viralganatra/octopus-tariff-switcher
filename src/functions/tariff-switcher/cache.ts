import type { AllProducts } from './schema';

let token: string;
let allProducts: AllProducts = [];

export function getCachedToken() {
  return token;
}

export function setCachedToken(newToken: string) {
  token = newToken;
  return token;
}

export function getCachedProducts() {
  return allProducts;
}

export function setCachedProducts(products: AllProducts) {
  allProducts = products;
  return allProducts;
}
