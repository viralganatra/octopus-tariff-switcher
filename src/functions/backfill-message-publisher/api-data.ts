import { isWithinInterval, parseISO } from 'date-fns';
import type { IsoDate } from '../../types/misc';
import type { ConsumptionIntervals, EletricityAgreements } from './schema';
import { chunkArray, processBatches } from '../../utils/batch';
import { TARIFFS, type TariffDisplayName } from '../../constants/tariff';
import { retryWithExponentialBackoff } from '../../utils/fetch';
import { fetchConsumption, fetchStandingCharge } from './queries';
import { fetchUnitRatesByTariff } from '../tariff-switcher/queries';
import type { TariffUnitRates } from '../tariff-switcher/schema';

export type TariffData = {
  isoDate: IsoDate;
  tariffCode: string;
  tariffName: TariffDisplayName;
  productCode: string;
  standingCharge: number;
  unitRates: TariffUnitRates;
  consumption: ConsumptionIntervals;
};

type ItemCacheBuilder = Pick<TariffData, 'isoDate' | 'tariffCode' | 'tariffName' | 'productCode'> &
  Partial<Omit<TariffData, 'isoDate' | 'tariffCode' | 'tariffName' | 'productCode'>>;

const BATCH_SIZE = 25;

export function findMatchingTariffForDate({
  pastTariffs,
  isoDate,
}: {
  pastTariffs: EletricityAgreements;
  isoDate: IsoDate;
}) {
  const targetDate = parseISO(isoDate);

  // Normalize target to start of day UTC
  const targetDateUtc = new Date(
    Date.UTC(targetDate.getUTCFullYear(), targetDate.getUTCMonth(), targetDate.getUTCDate()),
  );

  const matchingTariff = pastTariffs.find(({ validFrom, validTo }) => {
    const from = parseISO(validFrom);
    const to = parseISO(validTo);

    // isWithinInterval treats end as inclusive so we need to make it exclusive
    return (
      isWithinInterval(targetDateUtc, {
        start: from,
        end: to,
      }) && targetDateUtc.getTime() !== to.getTime()
    );
  });

  if (!matchingTariff) {
    throw new Error(`Unable to find matching tariff for date: ${isoDate}`);
  }

  return matchingTariff;
}

// Find the tariff code is in the format E-1R-COSY-22-12-08-A, extract COSY-22-12-08
// This isn't ideal, but there is no API endpoint that allows us to get a product code
// from a tariff code. Using the products endpoint with an available_at param doesn't work
// as you can be on an older tariff for a given date
function findProductCodeByTariffCode(tariffCode: string) {
  const regex = /(?:E-\dR-)(.*?)(?:-[A-Z])?$/;

  const parts = regex.exec(tariffCode);

  if (!parts || !parts[1]) {
    throw new Error(`No product code found from tariff code: ${tariffCode}`);
  }

  return parts[1];
}

export async function enrichDatesWithTariffData({
  dates,
  pastTariffs,
  mpan,
  serialNumber,
}: { dates: IsoDate[]; pastTariffs: EletricityAgreements; mpan: string; serialNumber: string }) {
  const itemCacheBuilder = new Map<IsoDate, ItemCacheBuilder>();
  const batchesOfDates = chunkArray(dates, BATCH_SIZE);

  // Build initial cache directly
  for (const isoDate of dates) {
    const { tariffCode } = findMatchingTariffForDate({ pastTariffs, isoDate });
    const productCode = findProductCodeByTariffCode(tariffCode);

    const tariff = TARIFFS.find(({ tariffCodeMatcher }) => tariffCode.includes(tariffCodeMatcher));

    if (!tariff) {
      throw new Error(`No matching tariff for: ${tariffCode}`);
    }

    itemCacheBuilder.set(isoDate, {
      isoDate,
      tariffCode,
      productCode,
      tariffName: tariff.displayName,
    });
  }

  // Fetch standing charge, unit rates & consumption in parallel batches
  const unitAndStandingChargesForDates = await processBatches({
    batches: batchesOfDates,
    processBatchItem: (isoDate) =>
      retryWithExponentialBackoff(async () => {
        const item = itemCacheBuilder.get(isoDate);

        if (!item || !item.productCode) {
          throw new Error(`Missing item or productCode for date: ${isoDate}`);
        }

        const { tariffCode, productCode } = item;

        const [standingCharge, unitRates, consumption] = await Promise.all([
          fetchStandingCharge({ isoDate, tariffCode, productCode }),
          fetchUnitRatesByTariff({ isoDate, tariffCode, productCode }),
          fetchConsumption({ isoDate, mpan, serialNumber }),
        ]);

        return { isoDate, standingCharge, unitRates, consumption };
      }),
  });

  // Populate item cache with standing charge, unit rates & consumption
  const itemCache = new Map<IsoDate, TariffData>();

  for (const {
    isoDate,
    standingCharge,
    unitRates,
    consumption,
  } of unitAndStandingChargesForDates) {
    const item = itemCacheBuilder.get(isoDate);

    if (!item || !item.productCode) {
      throw new Error(`Missing cache item for date: ${isoDate}`);
    }

    itemCache.set(isoDate, {
      isoDate,
      standingCharge,
      unitRates,
      consumption,
      tariffCode: item.tariffCode,
      tariffName: item.tariffName,
      productCode: item.productCode,
    });
  }

  return itemCache;
}
