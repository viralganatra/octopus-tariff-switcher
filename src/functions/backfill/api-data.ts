import { isWithinInterval, parseISO } from 'date-fns';
import type { IsoDate } from '../../types/misc';
import type { ConsumptionIntervals, EletricityAgreements } from './schema';
import { chunkArray, processBatches } from '../../utils/batch';
import { TARIFFS, type TariffDisplayName } from '../../constants/tariff';
import { getProductByTariff } from '../tariff-switcher/api-data';
import { retryWithExponentialBackoff } from '../../utils/fetch';
import { fetchConsumption, fetchStandingCharge } from './queries';
import { fetchUnitRatesByTariff } from '../tariff-switcher/queries';
import type { TariffUnitRates } from '../tariff-switcher/schema';

type ItemCache = {
  isoDate: IsoDate;
  tariffCode: string;
  tariffName: TariffDisplayName;
  productCode: string;
  standingCharge: number;
  unitRates: TariffUnitRates;
  consumption: ConsumptionIntervals;
};

type ItemCacheBuilder = Pick<ItemCache, 'isoDate' | 'tariffCode' | 'tariffName'> &
  Partial<Omit<ItemCache, 'isoDate' | 'tariffCode' | 'tariffName'>>;

const BATCH_SIZE = 25;

export function findMatchingTariffForDate({
  pastTariffs,
  isoDate,
}: {
  pastTariffs: EletricityAgreements;
  isoDate: IsoDate;
}) {
  const targetDate = parseISO(isoDate);

  // Normalize target to start of day UTC
  const targetDateUtc = new Date(
    Date.UTC(targetDate.getUTCFullYear(), targetDate.getUTCMonth(), targetDate.getUTCDate()),
  );

  const matchingTariff = pastTariffs.find(({ validFrom, validTo }) => {
    const from = parseISO(validFrom);
    const to = parseISO(validTo);

    // isWithinInterval treats end as inclusive so we need to make it exclusive
    return (
      isWithinInterval(targetDateUtc, {
        start: from,
        end: to,
      }) && targetDateUtc.getTime() !== to.getTime()
    );
  });

  if (!matchingTariff) {
    throw new Error(`Unable to find matching tariff for date: ${isoDate}`);
  }

  return matchingTariff;
}

export async function enrichDatesWithTariffData({
  dates,
  pastTariffs,
  mpan,
  serialNumber,
}: { dates: IsoDate[]; pastTariffs: EletricityAgreements; mpan: string; serialNumber: string }) {
  const itemCacheBuilder = new Map<IsoDate, ItemCacheBuilder>();
  const batchesOfDates = chunkArray(dates, BATCH_SIZE);

  // Build initial cache directly
  for (const isoDate of dates) {
    const { tariffCode } = findMatchingTariffForDate({ pastTariffs, isoDate });

    const tariff = TARIFFS.find(({ tariffCodeMatcher }) => tariffCode.includes(tariffCodeMatcher));

    if (!tariff) {
      throw new Error(`No matching tariff for: ${tariffCode}`);
    }

    itemCacheBuilder.set(isoDate, {
      isoDate,
      tariffCode,
      tariffName: tariff.displayName,
    });
  }

  // Fetch product codes for unique tariff names
  const uniqueTariffNames = new Set<ItemCache['tariffName']>();

  for (const item of itemCacheBuilder.values()) {
    uniqueTariffNames.add(item.tariffName);
  }

  const tariffToProduct = new Map<ItemCache['tariffName'], string>();

  await Promise.all(
    [...uniqueTariffNames].map(async (tariffName) => {
      const { code } = await getProductByTariff(tariffName);

      tariffToProduct.set(tariffName, code);
    }),
  );

  // Add product codes to cache
  for (const [isoDate, item] of itemCacheBuilder) {
    const productCode = tariffToProduct.get(item.tariffName);

    itemCacheBuilder.set(isoDate, { ...item, productCode });
  }

  // Fetch standing charge, unit rates & consumption in parallel batches
  const unitAndStandingChargesForDates = await processBatches({
    batches: batchesOfDates,
    processBatchItem: (isoDate) =>
      retryWithExponentialBackoff(async () => {
        const item = itemCacheBuilder.get(isoDate);

        if (!item || !item.productCode) {
          throw new Error(`Missing item or productCode for date: ${isoDate}`);
        }

        const { tariffCode, productCode } = item;

        const [standingCharge, unitRates, consumption] = await Promise.all([
          fetchStandingCharge({ isoDate, tariffCode, productCode }),
          fetchUnitRatesByTariff({ isoDate, tariffCode, productCode }),
          fetchConsumption({ isoDate, mpan, serialNumber }),
        ]);

        return { isoDate, standingCharge, unitRates, consumption };
      }),
  });

  // Populate item cache with standing charge, unit rates & consumption
  const itemCache = new Map<IsoDate, ItemCache>();

  for (const {
    isoDate,
    standingCharge,
    unitRates,
    consumption,
  } of unitAndStandingChargesForDates) {
    const item = itemCacheBuilder.get(isoDate);

    if (!item || !item.productCode) {
      throw new Error(`Missing cache item for date: ${isoDate}`);
    }

    itemCache.set(isoDate, {
      isoDate,
      standingCharge,
      unitRates,
      consumption,
      tariffCode: item.tariffCode,
      tariffName: item.tariffName,
      productCode: item.productCode,
    });
  }

  return itemCache;
}
