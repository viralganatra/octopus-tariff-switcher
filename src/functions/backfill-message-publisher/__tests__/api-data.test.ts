import type { IsoDate } from '../../../types/misc';
import type { EletricityAgreements } from '../schema';
import { enrichDatesWithTariffData, findMatchingTariffForDate } from '../api-data';
import { server } from '../../../mocks/node';
import { http, HttpResponse } from 'msw';
import {
  consumptionAgileFixture,
  standingChargeAgileFixture,
  unitRatesFixture,
} from '../../../mocks/fixtures';
import { setCachedProducts } from '../../tariff-switcher/cache';

describe('Finding matching tariffs', () => {
  it('should find the correct tariff for a given date', () => {
    const pastTariffs = [
      {
        validFrom: '2025-01-01T00:00:00Z',
        validTo: '2025-03-01T00:00:00Z',
        tariffCode: 'TARIFF-A',
      },
      {
        validFrom: '2025-03-01T00:00:00Z',
        validTo: '2025-06-01T00:00:00Z',
        tariffCode: 'TARIFF-B',
      },
      {
        validFrom: '2025-06-01T00:00:00Z',
        validTo: '2025-12-31T00:00:00Z',
        tariffCode: 'TARIFF-C',
      },
    ] as EletricityAgreements;

    const result = findMatchingTariffForDate({
      pastTariffs,
      isoDate: '2025-04-15' as IsoDate,
    });

    expect(result).toEqual({
      validFrom: '2025-03-01T00:00:00Z',
      validTo: '2025-06-01T00:00:00Z',
      tariffCode: 'TARIFF-B',
    });
  });

  it('should throw an error when no matching tariff is found', () => {
    const pastTariffs = [
      {
        validFrom: '2025-01-01T00:00:00Z',
        validTo: '2025-03-01T00:00:00Z',
        tariffCode: 'TARIFF-A',
      },
    ] as EletricityAgreements;

    expect(() => {
      findMatchingTariffForDate({
        pastTariffs,
        isoDate: '2025-04-15' as IsoDate,
      });
    }).toThrowError('Unable to find matching tariff for date: 2025-04-15');
  });

  it('should handle boundary dates correctly (exclusive end)', () => {
    const pastTariffs = [
      {
        validFrom: '2025-01-01T00:00:00Z',
        validTo: '2025-03-01T00:00:00Z',
        tariffCode: 'TARIFF-A',
      },
      {
        validFrom: '2025-03-01T00:00:00Z',
        validTo: '2025-06-01T00:00:00Z',
        tariffCode: 'TARIFF-B',
      },
    ] as EletricityAgreements;

    // Start of period should be included
    const resultStart = findMatchingTariffForDate({
      pastTariffs,
      isoDate: '2025-03-01' as IsoDate,
    });
    expect(resultStart.tariffCode).toBe('TARIFF-B');

    // End of period should be excluded
    expect(() => {
      findMatchingTariffForDate({
        pastTariffs,
        isoDate: '2025-06-01' as IsoDate,
      });
    }).toThrow();
  });
});

describe('Enrich dates with tariff data', () => {
  beforeEach(() => {
    setCachedProducts([]);
  });

  const pastTariffs = [
    {
      validFrom: '2020-01-01T00:00:00Z',
      validTo: '2024-03-01T00:00:00Z',
      tariffCode: 'E-1R-COSY-22-12-08-A',
    },
    {
      validFrom: '2025-03-01T00:00:00Z',
      validTo: '2025-06-01T00:00:00Z',
      tariffCode: 'E-1R-AGILE-24-10-01-A',
    },
  ] as EletricityAgreements;
  const dates = ['2020-02-15', '2025-04-15'] as IsoDate[];
  const mpan = '123456789';
  const serialNumber = 'SN12345';

  it('should enrich dates with tariff data', async () => {
    const data = await enrichDatesWithTariffData({ dates, pastTariffs, mpan, serialNumber });

    expect(data.size).toBe(2);

    const data2020 = data.get('2020-02-15' as IsoDate);
    expect(data2020?.isoDate).toBe('2020-02-15');
    expect(data2020?.productCode).toBe('COSY-22-12-08');
    expect(data2020?.tariffCode).toBe('E-1R-COSY-22-12-08-A');
    expect(data2020?.standingCharge).toBe(147.6062);
    expect(data2020?.tariffName).toBe('Cosy Octopus');
    expect(data2020?.consumption).matchSnapshot('Cosy consumption');
    expect(data2020?.unitRates).matchSnapshot('Cosy unit rates');

    const data2025 = data.get('2025-04-15' as IsoDate);
    expect(data2025?.isoDate).toBe('2025-04-15');
    expect(data2025?.productCode).toBe('AGILE-24-10-01');
    expect(data2025?.tariffCode).toBe('E-1R-AGILE-24-10-01-A');
    expect(data2025?.standingCharge).toBe(47.6062);
    expect(data2025?.tariffName).toBe('Agile Octopus');
    expect(data2025?.consumption).matchSnapshot('Agile consumption');
    expect(data2025?.unitRates).matchSnapshot('Agile unit rates');
  });

  it('should throw error when tariff is not found', async () => {
    const invalidPastTariffs = [
      {
        validFrom: '2020-01-01T00:00:00Z',
        validTo: '2024-03-01T00:00:00Z',
        tariffCode: 'E-1R-TEST-22-12-08-A',
      },
      {
        validFrom: '2025-03-01T00:00:00Z',
        validTo: '2025-06-01T00:00:00Z',
        tariffCode: 'E-1R-FOO-24-10-01-A',
      },
    ] as EletricityAgreements;

    await expect(
      enrichDatesWithTariffData({ dates, mpan, serialNumber, pastTariffs: invalidPastTariffs }),
    ).rejects.toThrowError('No matching tariff for: E-1R-TEST-22-12-08-A');
  });

  it('should handle batching for multiple dates', async () => {
    // Generate dates to test batching
    const manyDates = Array.from(
      { length: 30 },
      (_, i) => `2025-03-${String(i + 1).padStart(2, '0')}T00:00:00Z`,
    ) as IsoDate[];

    // Track request counts
    let standingChargeRequests = 0;
    let unitRateRequests = 0;
    let consumptionRequests = 0;

    server.use(
      http.get(
        'https://api.octopus.energy/v1/products/:productCode/electricity-tariffs/:tariffCode/standard-unit-rates',
        () => {
          unitRateRequests += 1;
          return HttpResponse.json(unitRatesFixture());
        },
      ),
      http.get(
        'https://api.octopus.energy/v1/products/:productCode/electricity-tariffs/:tariffCode/standing-charges/',
        () => {
          standingChargeRequests += 1;
          return HttpResponse.json(standingChargeAgileFixture);
        },
      ),
      http.get(
        'https://api.octopus.energy/v1/electricity-meter-points/:mpan/meters/:serialNumber/consumption/',
        () => {
          consumptionRequests += 1;
          return HttpResponse.json(consumptionAgileFixture());
        },
      ),
    );

    const promise = enrichDatesWithTariffData({
      dates: manyDates,
      pastTariffs: [
        {
          validFrom: '2025-01-01T00:00:00Z',
          validTo: '2026-12-01T00:00:00Z',
          tariffCode: 'E-1R-AGILE-24-10-01-A',
        },
      ],
      mpan,
      serialNumber,
    });

    await vi.runAllTimersAsync();

    const result = await promise;

    expect(result.size).toBe(30);
    expect(standingChargeRequests).toBe(30);
    expect(unitRateRequests).toBe(30);
    expect(consumptionRequests).toBe(30);
  });
});
