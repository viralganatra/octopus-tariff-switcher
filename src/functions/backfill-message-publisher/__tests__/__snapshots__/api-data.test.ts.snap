// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Enrich dates with tariff data > should enrich dates with tariff data > Agile consumption 1`] = `
[
  {
    "consumptionDelta": 74,
    "readAt": "2025-04-15T00:00:00Z",
    "readAtMs": 1744675200000,
  },
  {
    "consumptionDelta": 62,
    "readAt": "2025-04-15T00:30:00Z",
    "readAtMs": 1744677000000,
  },
  {
    "consumptionDelta": 77,
    "readAt": "2025-04-15T01:00:00Z",
    "readAtMs": 1744678800000,
  },
  {
    "consumptionDelta": 58,
    "readAt": "2025-04-15T01:30:00Z",
    "readAtMs": 1744680600000,
  },
  {
    "consumptionDelta": 77,
    "readAt": "2025-04-15T02:00:00Z",
    "readAtMs": 1744682400000,
  },
  {
    "consumptionDelta": 59,
    "readAt": "2025-04-15T02:30:00Z",
    "readAtMs": 1744684200000,
  },
  {
    "consumptionDelta": 74,
    "readAt": "2025-04-15T03:00:00Z",
    "readAtMs": 1744686000000,
  },
  {
    "consumptionDelta": 64,
    "readAt": "2025-04-15T03:30:00Z",
    "readAtMs": 1744687800000,
  },
  {
    "consumptionDelta": 69,
    "readAt": "2025-04-15T04:00:00Z",
    "readAtMs": 1744689600000,
  },
  {
    "consumptionDelta": 68,
    "readAt": "2025-04-15T04:30:00Z",
    "readAtMs": 1744691400000,
  },
  {
    "consumptionDelta": 63,
    "readAt": "2025-04-15T05:00:00Z",
    "readAtMs": 1744693200000,
  },
  {
    "consumptionDelta": 516,
    "readAt": "2025-04-15T05:30:00Z",
    "readAtMs": 1744695000000,
  },
  {
    "consumptionDelta": 303,
    "readAt": "2025-04-15T06:00:00Z",
    "readAtMs": 1744696800000,
  },
  {
    "consumptionDelta": 134,
    "readAt": "2025-04-15T06:30:00Z",
    "readAtMs": 1744698600000,
  },
  {
    "consumptionDelta": 83,
    "readAt": "2025-04-15T07:00:00Z",
    "readAtMs": 1744700400000,
  },
  {
    "consumptionDelta": 130,
    "readAt": "2025-04-15T07:30:00Z",
    "readAtMs": 1744702200000,
  },
  {
    "consumptionDelta": 112,
    "readAt": "2025-04-15T08:00:00Z",
    "readAtMs": 1744704000000,
  },
  {
    "consumptionDelta": 119,
    "readAt": "2025-04-15T08:30:00Z",
    "readAtMs": 1744705800000,
  },
  {
    "consumptionDelta": 103,
    "readAt": "2025-04-15T09:00:00Z",
    "readAtMs": 1744707600000,
  },
  {
    "consumptionDelta": 113,
    "readAt": "2025-04-15T09:30:00Z",
    "readAtMs": 1744709400000,
  },
  {
    "consumptionDelta": 147,
    "readAt": "2025-04-15T10:00:00Z",
    "readAtMs": 1744711200000,
  },
  {
    "consumptionDelta": 114,
    "readAt": "2025-04-15T10:30:00Z",
    "readAtMs": 1744713000000,
  },
  {
    "consumptionDelta": 100,
    "readAt": "2025-04-15T11:00:00Z",
    "readAtMs": 1744714800000,
  },
  {
    "consumptionDelta": 87,
    "readAt": "2025-04-15T11:30:00Z",
    "readAtMs": 1744716600000,
  },
  {
    "consumptionDelta": 156,
    "readAt": "2025-04-15T12:00:00Z",
    "readAtMs": 1744718400000,
  },
  {
    "consumptionDelta": 118,
    "readAt": "2025-04-15T12:30:00Z",
    "readAtMs": 1744720200000,
  },
  {
    "consumptionDelta": 175,
    "readAt": "2025-04-15T13:00:00Z",
    "readAtMs": 1744722000000,
  },
  {
    "consumptionDelta": 287,
    "readAt": "2025-04-15T13:30:00Z",
    "readAtMs": 1744723800000,
  },
  {
    "consumptionDelta": 795,
    "readAt": "2025-04-15T14:00:00Z",
    "readAtMs": 1744725600000,
  },
  {
    "consumptionDelta": 493,
    "readAt": "2025-04-15T14:30:00Z",
    "readAtMs": 1744727400000,
  },
  {
    "consumptionDelta": 108,
    "readAt": "2025-04-15T15:00:00Z",
    "readAtMs": 1744729200000,
  },
  {
    "consumptionDelta": 142,
    "readAt": "2025-04-15T15:30:00Z",
    "readAtMs": 1744731000000,
  },
  {
    "consumptionDelta": 126,
    "readAt": "2025-04-15T16:00:00Z",
    "readAtMs": 1744732800000,
  },
  {
    "consumptionDelta": 212,
    "readAt": "2025-04-15T16:30:00Z",
    "readAtMs": 1744734600000,
  },
  {
    "consumptionDelta": 179,
    "readAt": "2025-04-15T17:00:00Z",
    "readAtMs": 1744736400000,
  },
  {
    "consumptionDelta": 176,
    "readAt": "2025-04-15T17:30:00Z",
    "readAtMs": 1744738200000,
  },
  {
    "consumptionDelta": 299,
    "readAt": "2025-04-15T18:00:00Z",
    "readAtMs": 1744740000000,
  },
  {
    "consumptionDelta": 219,
    "readAt": "2025-04-15T18:30:00Z",
    "readAtMs": 1744741800000,
  },
  {
    "consumptionDelta": 197,
    "readAt": "2025-04-15T19:00:00Z",
    "readAtMs": 1744743600000,
  },
  {
    "consumptionDelta": 158,
    "readAt": "2025-04-15T19:30:00Z",
    "readAtMs": 1744745400000,
  },
  {
    "consumptionDelta": 152,
    "readAt": "2025-04-15T20:00:00Z",
    "readAtMs": 1744747200000,
  },
  {
    "consumptionDelta": 172,
    "readAt": "2025-04-15T20:30:00Z",
    "readAtMs": 1744749000000,
  },
  {
    "consumptionDelta": 163,
    "readAt": "2025-04-15T21:00:00Z",
    "readAtMs": 1744750800000,
  },
  {
    "consumptionDelta": 116,
    "readAt": "2025-04-15T21:30:00Z",
    "readAtMs": 1744752600000,
  },
  {
    "consumptionDelta": 88,
    "readAt": "2025-04-15T22:00:00Z",
    "readAtMs": 1744754400000,
  },
  {
    "consumptionDelta": 64,
    "readAt": "2025-04-15T22:30:00Z",
    "readAtMs": 1744756200000,
  },
  {
    "consumptionDelta": 77,
    "readAt": "2025-04-15T23:00:00Z",
    "readAtMs": 1744758000000,
  },
  {
    "consumptionDelta": 61,
    "readAt": "2025-04-15T23:30:00Z",
    "readAtMs": 1744759800000,
  },
]
`;

exports[`Enrich dates with tariff data > should enrich dates with tariff data > Agile unit rates 1`] = `
[
  {
    "unitCostInPence": 13.23168,
    "validFrom": "2025-04-15T22:00:00Z",
    "validFromMs": 1744754400000,
    "validTo": "2025-04-16T00:00:00Z",
    "validToMs": 1744761600000,
  },
  {
    "unitCostInPence": 26.98143,
    "validFrom": "2025-04-15T19:00:00Z",
    "validFromMs": 1744743600000,
    "validTo": "2025-04-15T22:00:00Z",
    "validToMs": 1744754400000,
  },
  {
    "unitCostInPence": 40.47225,
    "validFrom": "2025-04-15T16:00:00Z",
    "validFromMs": 1744732800000,
    "validTo": "2025-04-15T19:00:00Z",
    "validToMs": 1744743600000,
  },
  {
    "unitCostInPence": 13.23168,
    "validFrom": "2025-04-15T13:00:00Z",
    "validFromMs": 1744722000000,
    "validTo": "2025-04-15T16:00:00Z",
    "validToMs": 1744732800000,
  },
  {
    "unitCostInPence": 26.98143,
    "validFrom": "2025-04-15T07:00:00Z",
    "validFromMs": 1744700400000,
    "validTo": "2025-04-15T13:00:00Z",
    "validToMs": 1744722000000,
  },
  {
    "unitCostInPence": 13.23168,
    "validFrom": "2025-04-15T04:00:00Z",
    "validFromMs": 1744689600000,
    "validTo": "2025-04-15T07:00:00Z",
    "validToMs": 1744700400000,
  },
  {
    "unitCostInPence": 26.98143,
    "validFrom": "2025-04-15T00:00:00Z",
    "validFromMs": 1744675200000,
    "validTo": "2025-04-15T04:00:00Z",
    "validToMs": 1744689600000,
  },
]
`;

exports[`Enrich dates with tariff data > should enrich dates with tariff data > Cosy consumption 1`] = `
[
  {
    "consumptionDelta": 72,
    "readAt": "2020-02-15T01:00:00+01:00",
    "readAtMs": 1581724800000,
  },
  {
    "consumptionDelta": 897,
    "readAt": "2020-02-15T01:30:00+01:00",
    "readAtMs": 1581726600000,
  },
  {
    "consumptionDelta": 314,
    "readAt": "2020-02-15T02:00:00+01:00",
    "readAtMs": 1581728400000,
  },
  {
    "consumptionDelta": 328,
    "readAt": "2020-02-15T02:30:00+01:00",
    "readAtMs": 1581730200000,
  },
  {
    "consumptionDelta": 335,
    "readAt": "2020-02-15T03:00:00+01:00",
    "readAtMs": 1581732000000,
  },
  {
    "consumptionDelta": 314,
    "readAt": "2020-02-15T03:30:00+01:00",
    "readAtMs": 1581733800000,
  },
  {
    "consumptionDelta": 288,
    "readAt": "2020-02-15T04:00:00+01:00",
    "readAtMs": 1581735600000,
  },
  {
    "consumptionDelta": 308,
    "readAt": "2020-02-15T04:30:00+01:00",
    "readAtMs": 1581737400000,
  },
  {
    "consumptionDelta": 267,
    "readAt": "2020-02-15T05:00:00+01:00",
    "readAtMs": 1581739200000,
  },
  {
    "consumptionDelta": 302,
    "readAt": "2020-02-15T05:30:00+01:00",
    "readAtMs": 1581741000000,
  },
  {
    "consumptionDelta": 63,
    "readAt": "2020-02-15T06:00:00+01:00",
    "readAtMs": 1581742800000,
  },
  {
    "consumptionDelta": 95,
    "readAt": "2020-02-15T06:30:00+01:00",
    "readAtMs": 1581744600000,
  },
  {
    "consumptionDelta": 73,
    "readAt": "2020-02-15T07:00:00+01:00",
    "readAtMs": 1581746400000,
  },
  {
    "consumptionDelta": 108,
    "readAt": "2020-02-15T07:30:00+01:00",
    "readAtMs": 1581748200000,
  },
  {
    "consumptionDelta": 93,
    "readAt": "2020-02-15T08:00:00+01:00",
    "readAtMs": 1581750000000,
  },
  {
    "consumptionDelta": 159,
    "readAt": "2020-02-15T08:30:00+01:00",
    "readAtMs": 1581751800000,
  },
  {
    "consumptionDelta": 87,
    "readAt": "2020-02-15T09:00:00+01:00",
    "readAtMs": 1581753600000,
  },
  {
    "consumptionDelta": 80,
    "readAt": "2020-02-15T09:30:00+01:00",
    "readAtMs": 1581755400000,
  },
  {
    "consumptionDelta": 83,
    "readAt": "2020-02-15T10:00:00+01:00",
    "readAtMs": 1581757200000,
  },
  {
    "consumptionDelta": 62,
    "readAt": "2020-02-15T10:30:00+01:00",
    "readAtMs": 1581759000000,
  },
  {
    "consumptionDelta": 196,
    "readAt": "2020-02-15T11:00:00+01:00",
    "readAtMs": 1581760800000,
  },
  {
    "consumptionDelta": 152,
    "readAt": "2020-02-15T11:30:00+01:00",
    "readAtMs": 1581762600000,
  },
  {
    "consumptionDelta": 68,
    "readAt": "2020-02-15T12:00:00+01:00",
    "readAtMs": 1581764400000,
  },
  {
    "consumptionDelta": 80,
    "readAt": "2020-02-15T12:30:00+01:00",
    "readAtMs": 1581766200000,
  },
  {
    "consumptionDelta": 135,
    "readAt": "2020-02-15T13:00:00+01:00",
    "readAtMs": 1581768000000,
  },
  {
    "consumptionDelta": 80,
    "readAt": "2020-02-15T13:30:00+01:00",
    "readAtMs": 1581769800000,
  },
  {
    "consumptionDelta": 65,
    "readAt": "2020-02-15T14:00:00+01:00",
    "readAtMs": 1581771600000,
  },
  {
    "consumptionDelta": 77,
    "readAt": "2020-02-15T14:30:00+01:00",
    "readAtMs": 1581773400000,
  },
  {
    "consumptionDelta": 76,
    "readAt": "2020-02-15T15:00:00+01:00",
    "readAtMs": 1581775200000,
  },
  {
    "consumptionDelta": 145,
    "readAt": "2020-02-15T15:30:00+01:00",
    "readAtMs": 1581777000000,
  },
  {
    "consumptionDelta": 100,
    "readAt": "2020-02-15T16:00:00+01:00",
    "readAtMs": 1581778800000,
  },
  {
    "consumptionDelta": 161,
    "readAt": "2020-02-15T16:30:00+01:00",
    "readAtMs": 1581780600000,
  },
  {
    "consumptionDelta": 118,
    "readAt": "2020-02-15T17:00:00+01:00",
    "readAtMs": 1581782400000,
  },
  {
    "consumptionDelta": 128,
    "readAt": "2020-02-15T17:30:00+01:00",
    "readAtMs": 1581784200000,
  },
  {
    "consumptionDelta": 206,
    "readAt": "2020-02-15T18:00:00+01:00",
    "readAtMs": 1581786000000,
  },
  {
    "consumptionDelta": 137,
    "readAt": "2020-02-15T18:30:00+01:00",
    "readAtMs": 1581787800000,
  },
  {
    "consumptionDelta": 141,
    "readAt": "2020-02-15T19:00:00+01:00",
    "readAtMs": 1581789600000,
  },
  {
    "consumptionDelta": 135,
    "readAt": "2020-02-15T19:30:00+01:00",
    "readAtMs": 1581791400000,
  },
  {
    "consumptionDelta": 143,
    "readAt": "2020-02-15T20:00:00+01:00",
    "readAtMs": 1581793200000,
  },
  {
    "consumptionDelta": 160,
    "readAt": "2020-02-15T20:30:00+01:00",
    "readAtMs": 1581795000000,
  },
  {
    "consumptionDelta": 162,
    "readAt": "2020-02-15T21:00:00+01:00",
    "readAtMs": 1581796800000,
  },
  {
    "consumptionDelta": 124,
    "readAt": "2020-02-15T21:30:00+01:00",
    "readAtMs": 1581798600000,
  },
  {
    "consumptionDelta": 106,
    "readAt": "2020-02-15T22:00:00+01:00",
    "readAtMs": 1581800400000,
  },
  {
    "consumptionDelta": 69,
    "readAt": "2020-02-15T22:30:00+01:00",
    "readAtMs": 1581802200000,
  },
  {
    "consumptionDelta": 79,
    "readAt": "2020-02-15T23:00:00+01:00",
    "readAtMs": 1581804000000,
  },
  {
    "consumptionDelta": 58,
    "readAt": "2020-02-15T23:30:00+01:00",
    "readAtMs": 1581805800000,
  },
  {
    "consumptionDelta": 78,
    "readAt": "2020-02-16T00:00:00+01:00",
    "readAtMs": 1581807600000,
  },
  {
    "consumptionDelta": 56,
    "readAt": "2020-02-16T00:30:00+01:00",
    "readAtMs": 1581809400000,
  },
]
`;

exports[`Enrich dates with tariff data > should enrich dates with tariff data > Cosy unit rates 1`] = `
[
  {
    "unitCostInPence": 13.23168,
    "validFrom": "2020-03-03T22:00:00Z",
    "validFromMs": 1583272800000,
    "validTo": "2020-03-02T00:00:00Z",
    "validToMs": 1583107200000,
  },
]
`;
